Add hash_type MUHASH for gettxoutsetinfo

fjahr opened the pull request:
This is another Pr in the series PRs for Coinstatsindex (see overview in #18000). This PR adds the `hash_type` option `muhash` to `gettxoutsetinfo` through which the user can calculate the serialized muhash of the utxo set. This PR does not use the index yet.

fjahr committed:
rpc: Add hash_type MUHASH to gettxoutsetinfo

Also small style fix in rpc/util.cpp

fjahr commented:
Added another more extensive test, now ready for review.

Sjors commented:
Concept ACK. In the followup PR that adds a MuHash index, the RPC documentation should point out that this index can be used to dramatically speed up the RPC call when used with `muhash`.

fjahr commented:
@MarcoFalke I would like to hear your thoughts on 2109165e3b61d26a0496d5a1efd1c635a7f6a76a. I had it only in the test at first but put it into the framework because I think it could be useful for other tests as well and since it's just a method it does not affect the use of `P2PDataStore` in other tests. But the data it returns on each UTXO is kind of specific to my test.

MarcoFalke commented:
Seems fine to put it wherever you want, as long as reviewers are happy. I guess the decision mostly depends on whether this will be used by other tests.

fjahr commented:
rebased

fjahr commented:
Fixed a merge issue and pulled in latest changes from dependency PRs.

fjahr committed:
refactor: Separate hash and stats calculation in coinstats

fjahr commented:
Removed some code dublication

fjahr commented:
Rebased after #19105 was merged.

mjdietzx commented on test/functional/feature_utxo_set_hash.py:
```diff
@@ -0,0 +1,90 @@
+        conn.send_blocks_and_test([block2], node, success=True)
+
+        # Calculate the MuHash of the UTXO set estimate
+        muhash = MuHash3072()
```
Would this test be a little better if instead of forming the utxo set at the end, you maintained a running sum and added as blocks were mined, and then added/subtracted when UTXOs were spent and created in `tx1`, `tx2`?

fjahr replied:
I don't think so because we need to somehow keep track of what is really in the UTXO set and I think this would make test harder to follow if I did it this way. And I feel like this is more of the standard way of how we structure our tests.

fjahr committed:
test: Add test for gettxoutsetinfo RPC with MuHash

fjahr committed:
refactor: Improve encapsulation between MuHash3072 and Num3072

Also fixes a typo.

fjahr commented on src/Makefile.am:
```diff
@@ -397,6 +397,8 @@ crypto_libbitcoin_crypto_base_a_SOURCES = \
   crypto/poly1305.h \
   crypto/poly1305.cpp \
+  crypto/muhash.h \
+  crypto/muhash.cpp \
```
Sorry, this was part of a previous PR that this one is based on, so I don't want to touch this anymore.

fjahr commented on test/functional/feature_utxo_set_hash.py:
```diff
@@ -0,0 +1,90 @@
+
+            # Save first block because we want to spend its coinbase.
+            if i == 0:
+                block1 = block
```
It was and I did it because I thought there was a good chance that this code could be used for other tests as well and it also made some things a bit easier. But overall of course it adds complexity and I am lacking a specific example where it would be used so I have removed the P2P based implementation and simplified it in a similar way as you imagined, I think. Please check it out.

fjahr commented:
Rebased and addressed feedback @mjdietzx . Thanks for reviewing and sorry for the long wait, I was focussing on #19055 which is now merged and which makes this one my high prio. I also added a small refactor commit to improve encapsulation between `MuHash3072` and `Num3072`.

FelixWeis commented on src/node/coinstats.cpp:
```diff
@@ -24,31 +25,47 @@ static uint64_t GetBogoSize(const CScript& scriptPubKey)
 }
 
-static void ApplyStats(CCoinsStats& stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
+static void ApplyHash(CCoinsStats &stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs, std::map<uint32_t, Coin>::const_iterator it)
```
nit: `CCoinsStats& stats`

fjahr replied:
Done

fjahr commented:
> There is a deterministic regtest blockchain for the test `rpc_getblockstats.py` which could (additionally) test the finalized hash against constants, especially in the future when one tests for correct rolling behaviour between blocks.

Thanks, that's a good point but I think this will be even more interesting when we get to the actual index when the RPC can give hashes for specific heights so I think I will leave this for one of the follow-ups of this one.

fjahr commented:
Addressed Feedback by @FelixWeis and fixed the functional test.

sipa commented on src/node/coinstats.cpp:
```diff
@@ -42,8 +43,20 @@ static void ApplyHash(CCoinsStats &stats, CHashWriter& ss, const uint256& hash,
+    ss << outpoint;
+    ss << (uint32_t)(coin.nHeight * 2 + coin.fCoinBase);
+    ss << coin.out;
+    muhash.Insert(MakeUCharSpan(ss.str()));
```
In commit "rpc: Add hash_type MUHASH to gettxoutsetinfo"

I think extracting an `std::string` here is unnecessary. 

```c++
muhash.Insert(MakeUCharSpan(ss));
```

should work just as well, and avoid the copy.

fjahr replied:
Fixed

sipa commented on src/node/coinstats.cpp:
```diff
@@ -24,31 +24,35 @@ static uint64_t GetBogoSize(const CScript& scriptPubKey)
 }
 
-static void ApplyStats(CCoinsStats& stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
+static void ApplyHash(CCoinsStats &stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs, std::map<uint32_t, Coin>::const_iterator it)
```
In commit "refactor: Seperate hash and stats calculation in coinstats"

Nit: seperate -> separate in commit title

fjahr replied:
Fixed

fjahr commented on src/crypto/muhash.cpp:
```diff
@@ -276,18 +275,33 @@ void Num3072::Divide(const Num3072& a)
-    uint256 hashed_in = (CHashWriter(SER_DISK, 0) << in).GetSHA256();
-    unsigned char tmp[BYTE_SIZE];
-    ChaCha20(hashed_in.data(), hashed_in.size()).Keystream(tmp, BYTE_SIZE);
+Num3072::Num3072(const unsigned char data[BYTE_SIZE]) {
```
Done

jonatack commented on test/functional/rpc_blockchain.py:
```diff
@@ -267,6 +267,13 @@ def _test_gettxoutsetinfo(self):
         # hash_type none should not return a UTXO set hash.
         res5 = node.gettxoutsetinfo(hash_type='none')
         assert 'hash_serialized_2' not in res5
+        assert 'muhash' not in res5
```
bde3feba should `assert 'muhash' not in` be added to any of res-res4

fjahr replied:
Done.

jonatack commented on src/rpc/util.cpp:
```diff
@@ -117,16 +117,18 @@ CoinStatsHashType ParseHashType(const UniValue& param, const CoinStatsHashType d
+    } else if (hash_type_input == "muhash") {
+        return CoinStatsHashType::MUHASH;
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("%d is not a valid hash_type", hash_type_input));
```
bde3febaf this could perhaps be a bit simpler

<details><summary>diff</summary><p>

```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 576dabc349..3a07e5620d 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -1079,7 +1079,7 @@ static RPCHelpMan gettxoutsetinfo()
     CCoinsStats stats;
     ::ChainstateActive().ForceFlushStateToDisk();
 
-    const CoinStatsHashType hash_type = ParseHashType(request.params[0], CoinStatsHashType::HASH_SERIALIZED);
+    const CoinStatsHashType hash_type{request.params[0].isNull() ? CoinStatsHashType::HASH_SERIALIZED : ParseHashType(request.params[0].get_str())};
 
     CCoinsView* coins_view = WITH_LOCK(cs_main, return &ChainstateActive().CoinsDB());
     NodeContext& node = EnsureNodeContext(request.context);
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 30e09a9aac..d8dadf41f7 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -113,14 +113,8 @@ std::vector<unsigned char> ParseHexO(const UniValue& o, std::string strKey)
     return ParseHexV(find_value(o, strKey), strKey);
 }
 
-CoinStatsHashType ParseHashType(const UniValue& param, const CoinStatsHashType default_type)
+CoinStatsHashType ParseHashType(const std::string& hash_type_input)
 {
-    if (param.isNull()) {
-        return default_type;
-    }
-
-    std::string hash_type_input = param.get_str();
-
     if (hash_type_input == "hash_serialized_2") {
         return CoinStatsHashType::HASH_SERIALIZED;
     } else if (hash_type_input == "none") {
diff --git a/src/rpc/util.h b/src/rpc/util.h
index 942c243718..611d281014 100644
--- a/src/rpc/util.h
+++ b/src/rpc/util.h
@@ -77,7 +77,7 @@ extern uint256 ParseHashO(const UniValue& o, std::string strKey);
 extern std::vector<unsigned char> ParseHexV(const UniValue& v, std::string strName);
 extern std::vector<unsigned char> ParseHexO(const UniValue& o, std::string strKey);
 
-CoinStatsHashType ParseHashType(const UniValue& param, const CoinStatsHashType default_type);
+CoinStatsHashType ParseHashType(const std::string& hash_type_input);
 
 extern CAmount AmountFromValue(const UniValue& value);
 extern std::string HelpExampleCli(const std::string& methodname, const std::string& args);
```
</p></details>

jonatack replied:
Could also drop the header file declaration and move the function into the same file as the caller, unless you expect it to have callers in other files.

fjahr replied:
You're right, I don't foresee any other callers at the moment so I moved the function into `rpc/blockchain.cpp`.

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -1063,6 +1063,7 @@ static RPCHelpMan gettxoutsetinfo()
                         {RPCResult::Type::NUM, "txouts", "The number of unspent transaction outputs"},
                         {RPCResult::Type::NUM, "bogosize", "A meaningless metric for UTXO set size"},
+                        {RPCResult::Type::STR_HEX, "muhash", "The serialized hash (only present if 'muhash' hash_type is chosen)"},
                         {RPCResult::Type::STR_HEX, "hash_serialized_2", "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
```
bde3febaf pass as optional results
```diff
-                        {RPCResult::Type::STR_HEX, "muhash", "The serialized hash (only present if 'muhash' hash_type is chosen)"},
-                        {RPCResult::Type::STR_HEX, "hash_serialized_2", "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
+                        {RPCResult::Type::STR_HEX, "muhash", /* optional */ true, "The serialized hash (only present if 'muhash' hash_type is chosen)"},
+                        {RPCResult::Type::STR_HEX, "hash_serialized_2", /* optional */ true, "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
```

fjahr replied:
Done.

jonatack commented on src/node/coinstats.cpp:
```diff
@@ -24,31 +25,47 @@ static uint64_t GetBogoSize(const CScript& scriptPubKey)
+
+    CDataStream ss(SER_DISK, PROTOCOL_VERSION);
+    ss << outpoint;
+    ss << (uint32_t)(coin.nHeight * 2 + coin.fCoinBase);
```
bde3feba [named cast](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Res-casts-named)
```suggestion
    ss << static_cast<uint32_t>(coin.nHeight * 2 + coin.fCoinBase);
```

fjahr replied:
Done.

jonatack commented on test/functional/feature_utxo_set_hash.py:
```diff
@@ -0,0 +1,75 @@
+from test_framework.messages import (
+    COutPoint,
+    FromHex,
+    CBlock,
```
7a96f77c nit, sort

fjahr replied:
Done.

jonatack commented on test/functional/feature_utxo_set_hash.py:
```diff
@@ -0,0 +1,75 @@
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+import struct
```
7a96f77c nit, place standard library imports before application/library specific imports per https://www.python.org/dev/peps/pep-0008/#imports

fjahr replied:
Done.

jonatack reviewed (commented):
Built and tested at each commit. This seems close. A few suggestions below.

fjahr commented on src/crypto/muhash.h:
```diff
@@ -91,8 +94,6 @@ class Num3072
 class MuHash3072
```
Fixed

fjahr commented:
Addressed feedback by @jonatack

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -1047,13 +1047,26 @@ static RPCHelpMan pruneblockchain()
                 "Note this call may take some time.\n",
                 {
-                    {"hash_type", RPCArg::Type::STR, /* default */ "hash_serialized_2", "Which UTXO set hash should be calculated. Options: 'hash_serialized_2' (the legacy algorithm), 'none'."},
+                    {"hash_type", RPCArg::Type::STR, /* default */ "hash_serialized_2", "Which UTXO set hash should be calculated. Options: 'hash_serialized_2' (the legacy algorithm), 'muhash', 'none'."},
```
good catch in the latest push
```diff
- Options: 'hash_serialized_2' (the legacy algorithm), 'none'."},
+ Options: 'hash_serialized_2' (the legacy algorithm), 'muhash', 'none'."},
```

fjahr replied:
Yeah, I feel like I fixed this before but I guess I messed up a rebase somewhere ;)

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -1047,13 +1047,26 @@ static RPCHelpMan pruneblockchain()
+    } else if (hash_type_input == "muhash") {
+        return CoinStatsHashType::MUHASH;
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("%d is not a valid hash_type", hash_type_input));
```
```suggestion
        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("%s is not a valid hash_type", hash_type_input));
```

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -1047,13 +1047,26 @@ static RPCHelpMan pruneblockchain()
+    } else if (hash_type_input == "muhash") {
+        return CoinStatsHashType::MUHASH;
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("%d is not a valid hash_type", hash_type_input));
```
726c9504 I overlooked this in my first review
```suggestion

        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("%s is not a valid hash_type", hash_type_input));
```

jonatack replied:
Perhaps add a functional test for this.

jonatack replied:
Oh, maybe it works anyway
```
./src/bitcoin-cli -signet gettxoutsetinfo a
error code: -8
error message:
a is not a valid hash_type
```

fjahr replied:
Done

jonatack commented on src/node/coinstats.cpp:
```diff
@@ -117,9 +138,17 @@ static void PrepareHash(CHashWriter& ss, const CCoinsStats& stats)
     ss << stats.hashBlock;
 }
 static void PrepareHash(std::nullptr_t, CCoinsStats& stats) {}
+// Muhash does not need the prepare step
```
726c9504 pico-nit, feel free to ignore, s/Muhash/MuHash/

fjahr replied:
Done

jonatack commented:
ACK 2e5617f7f4bcece4834852cc925068e47ceeffa3 code review, reviewed diff last review per `git diff 7a96f77 2e5617f`, tested rebased on current master

Happy to re-ACK for the changes above.

<details><summary>Some manual command line testing.</summary><p>

```
# bsi is an alias for ./src/bitcoin-cli -signet

$ ((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo
{
  "height": 21927,
  "bestblock": "0000000d69e0c1800fd7103a7f1aaf24ec7b8c1db92257bda1035dcfd8e71c30",
  "transactions": 17430,
  "txouts": 17550,
  "bogosize": 1263989,
  "hash_serialized_2": "c8dbf6ab5094c36935ca89fc8e347b65ab17320f0937c5ac942274a1a0027892",
  "disk_size": 1264123,
  "total_amount": 1096350.00000000
}

$ ((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo muhash
{
  "height": 21927,
  "bestblock": "0000000d69e0c1800fd7103a7f1aaf24ec7b8c1db92257bda1035dcfd8e71c30",
  "transactions": 17430,
  "txouts": 17550,
  "bogosize": 1263989,
  "muhash": "cd74b9cfe35dfee561ed730953ed15085e83c096db30d5c03a8ab1a6a137077a",
  "disk_size": 1264123,
  "total_amount": 1096350.00000000
}

$ ((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo none
{
  "height": 21927,
  "bestblock": "0000000d69e0c1800fd7103a7f1aaf24ec7b8c1db92257bda1035dcfd8e71c30",
  "transactions": 17430,
  "txouts": 17550,
  "bogosize": 1263989,
  "disk_size": 1264123,
  "total_amount": 1096350.00000000
}

$ ((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo non
error code: -8
error message:
non is not a valid hash_type
```
</p></details>

jonatack commented:
ACK 2e5617f7f4bcece4834852cc925068e47ceeffa3 code review, reviewed diff last review per `git diff 7a96f77 2e5617f`, tested rebased on current master

Happy to re-ACK for the changes above.

<details><summary>Some manual command line testing.</summary><p>

```
# bsi is an alias for ./src/bitcoin-cli -signet

((HEAD detached at origin/pr/19145))$ bsi help gettxoutsetinfo
gettxoutsetinfo ( "hash_type" )

Returns statistics about the unspent transaction output set.
Note this call may take some time.

Arguments:
1. hash_type    (string, optional, default=hash_serialized_2) Which UTXO set hash should be calculated. Options: 'hash_serialized_2' (the legacy algorithm), 'muhash', 'none'.

Result:
{                                 (json object)
  "height" : n,                   (numeric) The current block height (index)
  "bestblock" : "hex",            (string) The hash of the block at the tip of the chain
  "transactions" : n,             (numeric) The number of transactions with unspent outputs
  "txouts" : n,                   (numeric) The number of unspent transaction outputs
  "bogosize" : n,                 (numeric) A meaningless metric for UTXO set size
  "muhash" : "hex",               (string, optional) The serialized hash (only present if 'muhash' hash_type is chosen)
  "hash_serialized_2" : "hex",    (string, optional) The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)
  "disk_size" : n,                (numeric) The estimated size of the chainstate on disk
  "total_amount" : n              (numeric) The total amount
}

Examples:
> bitcoin-cli gettxoutsetinfo 
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "gettxoutsetinfo", "params": []}' -H 'content-type: text/plain;' http://127.0.0.1:8332/

((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo
{
  "height": 21927,
  "bestblock": "0000000d69e0c1800fd7103a7f1aaf24ec7b8c1db92257bda1035dcfd8e71c30",
  "transactions": 17430,
  "txouts": 17550,
  "bogosize": 1263989,
  "hash_serialized_2": "c8dbf6ab5094c36935ca89fc8e347b65ab17320f0937c5ac942274a1a0027892",
  "disk_size": 1264123,
  "total_amount": 1096350.00000000
}

((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo hash_serialized_2
{
  "height": 21928,
  "bestblock": "000000c3c7e614717b8cdf6c33de7a8a7d97a856eba6640b481a0029fc774249",
  "transactions": 17431,
  "txouts": 17551,
  "bogosize": 1264061,
  "hash_serialized_2": "3e31af57df9dac6bf8924560b647fc81f01bdac1c0fe4af2bc12b61c81d87d61",
  "disk_size": 1264123,
  "total_amount": 1096400.00000000
}

((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo muhash
{
  "height": 21927,
  "bestblock": "0000000d69e0c1800fd7103a7f1aaf24ec7b8c1db92257bda1035dcfd8e71c30",
  "transactions": 17430,
  "txouts": 17550,
  "bogosize": 1263989,
  "muhash": "cd74b9cfe35dfee561ed730953ed15085e83c096db30d5c03a8ab1a6a137077a",
  "disk_size": 1264123,
  "total_amount": 1096350.00000000
}

((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo none
{
  "height": 21927,
  "bestblock": "0000000d69e0c1800fd7103a7f1aaf24ec7b8c1db92257bda1035dcfd8e71c30",
  "transactions": 17430,
  "txouts": 17550,
  "bogosize": 1263989,
  "disk_size": 1264123,
  "total_amount": 1096350.00000000
}

((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo non
error code: -8
error message:
non is not a valid hash_type

((HEAD detached at origin/pr/19145))$ bsi gettxoutsetinfo MuHash
error code: -8
error message:
MuHash is not a valid hash_type
```
</p></details>

fjahr reviewed (commented):
Addressed further feedback by @jonatack . Thanks for reviewing!

jonatack commented:
This might need a release note (even if the note will be updated in planned follow-ups to this).

Sjors commented:
Muhash at height 668,047: `c263d20036cc983510e9b347cb88e947168dd5665ed00b0e6118d07a7704f810`

Sjors commented on src/node/coinstats.cpp:
```diff
@@ -24,7 +25,7 @@ static uint64_t GetBogoSize(const CScript& scriptPubKey)
 }
 
-static void ApplyHash(CCoinsStats &stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs, std::map<uint32_t, Coin>::const_iterator it)
+static void ApplyHash(CCoinsStats& stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs, std::map<uint32_t, Coin>::const_iterator it)
```
e6856a3809e54dc4de03cb4cd4f9d78495cc0625: nit this undoes the move of `&` in the previous commit

fjahr replied:
Fixed.

Sjors commented on src/node/coinstats.cpp:
```diff
@@ -111,6 +124,10 @@ bool GetUTXOStats(CCoinsView* view, CCoinsStats& stats, CoinStatsHashType hash_t
     case(CoinStatsHashType::NONE): {
         return GetUTXOStats(view, stats, nullptr, interruption_point);
     }
+    case(CoinStatsHashType::MUHASH): {
```
e6856a3809e54dc4de03cb4cd4f9d78495cc0625 nit: handle `MUHASH` before `NONE`? (same with `PrepareHash ` and `ParseHashType`)

fjahr replied:
Done.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -1063,7 +1076,8 @@ static RPCHelpMan gettxoutsetinfo()
                         {RPCResult::Type::NUM, "txouts", "The number of unspent transaction outputs"},
                         {RPCResult::Type::NUM, "bogosize", "A meaningless metric for UTXO set size"},
-                        {RPCResult::Type::STR_HEX, "hash_serialized_2", "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
+                        {RPCResult::Type::STR_HEX, "muhash", /* optional */ true, "The serialized hash (only present if 'muhash' hash_type is chosen)"},
```
e6856a3809e54dc4de03cb4cd4f9d78495cc0625 nit: move down?

fjahr replied:
Done.

Sjors commented on src/node/coinstats.cpp:
```diff
@@ -117,9 +138,17 @@ static void PrepareHash(CHashWriter& ss, const CCoinsStats& stats)
 static void FinalizeHash(CHashWriter& ss, CCoinsStats& stats)
 {
-    stats.hashSerialized = ss.GetHash();
+    stats.hashSerialized = ss.GetSHA256();
```
This changes the result of `gettxoutsetinfo hash_serialized_2`

fjahr replied:
Huh, no idea how that happened, makes no sense at all, especially inside that commit. Fixed.

Sjors reviewed (changes requested):
Reviewed 3506d9080166bc250e941d791e3e8a2b85d5bd6b, almost happy... 

Muhash at height 668,212: `d0192e6604c3dc9a78f47bfc5ae84b8c0f6baf960d7ba9305664bbb9f65ef092`.

`hash_serialized_2` is `63fa057a656168ed832d8abbec8008948fc17aecde9f346d69043871a746941f` on master, but `15d3629db2fcbed27a1fd9d35c15b5ac6f0a70f5723ebf9cf87e3df0b0feb061` here (`efe73b1f89e82d65100c73c06d9a01ec19cc7e029d93caaa8bfded4d1723d484` if I revert `GetHash()`)

You may want to hardcode a `hash_serialized_2` example in the test suite.

fjahr committed:
test: Add test for deterministic UTXO set hash results

fjahr commented:
> Reviewed [3506d90](https://github.com/bitcoin/bitcoin/commit/3506d9080166bc250e941d791e3e8a2b85d5bd6b), almost happy...
> 
> Muhash at height 668,212: `d0192e6604c3dc9a78f47bfc5ae84b8c0f6baf960d7ba9305664bbb9f65ef092`.
> 
> `hash_serialized_2` is `63fa057a656168ed832d8abbec8008948fc17aecde9f346d69043871a746941f` on master, but `15d3629db2fcbed27a1fd9d35c15b5ac6f0a70f5723ebf9cf87e3df0b0feb061` here (`efe73b1f89e82d65100c73c06d9a01ec19cc7e029d93caaa8bfded4d1723d484` if I revert `GetHash()`)
> 
> You may want to hardcode a `hash_serialized_2` example in the test suite.

Thanks a lot for testing @Sjors ! I should have addressed all your comments.

The difference in that hash was caused by using `outputs.end()` where I should have used `std::prev(outputs.end())` in `ApplyHash()`. I was actually sure we had some deterministic tests in place for that hash already but obviously, we didn't. So I added a new commit with a small test calculating the hash based on the cached chain to add such coverage.

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -1063,7 +1076,8 @@ static RPCHelpMan gettxoutsetinfo()
                         {RPCResult::Type::NUM, "bogosize", "A meaningless metric for UTXO set size"},
-                        {RPCResult::Type::STR_HEX, "hash_serialized_2", "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
+                        {RPCResult::Type::STR_HEX, "hash_serialized_2", /* optional */ true, "The serialized hash (only present if 'hash_serialized_2' hash_type is chosen)"},
+                        {RPCResult::Type::STR_HEX, "muhash", /* optional */ true, "The serialized hash (only present if 'muhash' hash_type is chosen)"},
```
nit, in these two lines, could replace the single quotes with double ones for consistency, e.g. `s/'/\"/`

fjahr replied:
Yes, will do if I have to retouch! Thanks for re-reviewing!

jonatack reviewed (commented):
Good catch and nice improvements.

Tested re-ACK e987ae5a554c9952812746c29f2766bacea4b727 per `git diff 3506d90 e987ae5`, reviewed diff, debug built, ran gettxoutsetinfo -signet and help on this branch vs master, at height 23127 both returned `hash_serialized_2` of `2b72d65f3b6efb2311f58374ea2b939abf49684d44f4bafda45faa3b5452a454` and this branch returned `muhash` of `c9f1ff12d345ccf9939c6bbf087e6f7399b6115adee1569287e9c5c43dbb475c`

Only if you have to retouch and if you agree, one nit below (and maybe use single/double quotes consistently in the functional tests).

Sjors commented:
tACK e987ae5

I get the same MuHash as before for block 668,212 and `hash_serialized_2` now matches what I found on master.

achow101 commented:
ACK e987ae5a554c9952812746c29f2766bacea4b727

It would be nice if the test used MiniWallet or otherwise didn't rely on the wallet, but it's fine for now.

ryanofsky commented on src/node/coinstats.cpp:
```diff
@@ -24,31 +24,35 @@ static uint64_t GetBogoSize(const CScript& scriptPubKey)
 }
 
-static void ApplyStats(CCoinsStats& stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
+static void ApplyHash(CCoinsStats& stats, CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs, std::map<uint32_t, Coin>::const_iterator it)
```
In commit "refactor: Separate hash and stats calculation in coinstats" (2474645f3b15687e7f196b89eb935d6e6a98a9da)

`stats` arguments in unused in all three overloads and could be dropped.

Also, passing iterators around like this seems unnecessarily elaborate. It makes sense to separate stats and hash computations, but is there a reason both computations need to share the a single `for` loop over the output map? I would think you could make all the functions separate like:

static void ApplyHash(CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs);
static void ApplyHash(MuHash3072& muhash, const uint256& hash, const std::map<uint32_t, Coin>& outputs);
static void ApplyHash(std::nullptr_t, const uint256& hash, const std::map<uint32_t, Coin>& outputs);
static void ApplyStats(CCoinsStats& stats, const uint256& hash, const std::map<uint32_t, Coin>& outputs);

ryanofsky reviewed (approved):
Code review ACK e987ae5a554c9952812746c29f2766bacea4b727. Looks very good. I left one suggestion to simplify code, but feel free to ignore it here and maybe consider it for later since PR has already had a lot of review.

fjahr commented:
Thanks a lot for the reviews @achow101 and @ryanofsky ! I will add your suggestions if I have to push again, otherwise I will put them into the next PR in this series.

MarcoFalke commented on test/functional/feature_utxo_set_hash.py:
```diff
@@ -0,0 +1,86 @@
+    def test_deterministic_hash_results(self):
+        self.log.info("Test deterministic UTXO set hash results")
+
+        # These depend on the setup_clean_chain option, the chain loaded from the cache
```
there is no loaded chain and the utxo set is empty. Which is also useful to check, but doesn't check if the hash is deterministic?


Did you forget to load the chain from `./test/functional/data/rpc_getblockstats.json`?
